ballerina/googlesheets_client.bal

```diff
@@ -0,0 +1,520 @@
+
+        do {
+            return check 'object.cloneWithType(rowType);
+        } on fail error e {
+            return <Error>e;
+        }
+    }
+
+    # + rowType - The type description of the entity to be retrieved
+    # + typeMap - The data types of the record
+    # + fields - The fields to be retrieved
+    # + include - The associations to be retrieved
+    # + return - A stream of records in the `rowType` type or a `persist:Error` if the operation fails
+    public isolated function runReadQuery(typedesc<record {}> rowType, map<anydata> typeMap, string[] fields = [], string[] include = [])
+    returns stream<record {}, error?>|Error {
+        return self.query(self.addKeyFields(fields));
+    }
+
+    # + rowType - The type description of the entity to be retrieved
+    # + typeMap - The data types of the record
+    # + fields - The fields to be retrieved
+    # + include - The associations to be retrieved
+    # + return - A stream of records in the `rowType` type or a `persist:Error` if the operation fails
+    public isolated function readTableAsStream(typedesc<record {}> rowType, map<anydata> typeMap, string[] fields = [], string[] include = []) returns stream<record {}, Error?>|Error {
+        string query = "select *";
+        string|error encodedQuery = url:encode(query, "UTF-8");
+        if encodedQuery is error {
+            return <Error>error(encodedQuery.message());
+        }
+        http:QueryParams queries = {"gid": self.sheetId, "range": self.range, "tq": <string>encodedQuery, "tqx": "out:csv"};
+        http:Response|error response = self.httpClient->/d/[self.spreadsheetId]/gviz/tq(params = queries);
+        if response is error {
+            return <Error>error(response.message());
+        }
+        string|error textResponse = response.getTextPayload();
+        if (textResponse !is error) {
+            string[] responseRows = re `\n`.split(textResponse);
+            record {}[] rowTable = [];
+            if responseRows.length() == 0 {
+                return <Error>error("Error: the spreadsheet is not initialised correctly. ");
+            } else if responseRows.length() == 1 {
+                return rowTable.toStream();
+            }
+            string[] columnNames = re `,`.split(responseRows[0]);
+            foreach string rowString in responseRows.slice(1) {
+                int i = 0;
+                record {} rowArray = {};
+                string[] rowValues = re `,`.split(rowString);
+                foreach string rowValue in rowValues {
+                    string columnName = re `"`.replaceAll(columnNames[i], "");
+                    string value = re `"`.replaceAll(rowValue, "");
+                    string dataType = self.dataTypes.get(columnName).toString();
+                    if (dataType == "int") {
+                        (string|int|decimal|time:Date|time:TimeOfDay|time:Civil|time:Utc)|error typedValue = self.dataConverter(value, dataType);
```

Comment:
Can consider defining a new type simialr to `sql:Value`
https://github.com/ballerina-platform/module-ballerina-sql/blob/master/ballerina/types.bal#LL36C101-L36C111

```ballerina
public type TypedValue string|int|decimal|time:Date|time:TimeOfDay|time:Civil|time:Utc;

TypedValue|error = self.dataConverter(value, dataType);
```

Then you can use this instead of defining the type every time

Reply:
updated in [4346c39](https://github.com/ballerina-platform/module-ballerina-persist/pull/201/commits/4346c3959a465aeb05de0872b94ea1056a715a9a)